app/components/zig_codeblocks.py

```diff
@@ -0,0 +1,327 @@
+        # From Qwerasd:
+        #   Oh is it a safeguard against catastrophic backtracking?
+        #   [...]
+        #   I got distracted and checked the Discord source and this is the logic,
+        #   highlighting is disabled under these circumstances:
+        #   * The src contains 15 or more consecutive slashes
+        #   * The src contains a line with 1000 or more characters
+        #   * The src contains more than 30 slashes with anything in between as long as
+        #     it's not a line that isn't in the form ^\s*\/\/.* and contains a character
+        #     other than /
+        # These replace calls are an attempt to work around these limitations.
+        .replace("///", "\x1b[0m///")
+        .replace("// ", "\x1b[0m// ")
+    )
+
+
+class ZigCodeblockActions(discord.ui.View):
+    def __init__(self, message: discord.Message) -> None:
+        super().__init__()
+        self._message = message
+        self._replaced_message_content = custom_process_markdown(message.content)
+        self.replace.disabled = (
+            len(message_to_codeblocks[message]) > 1
+            or len(self._replaced_message_content) > 2000
+            or len(message.attachments) > 0
+        )
+
+    @discord.ui.button(
+        label="Dismiss",
+        emoji="❌",
+        style=discord.ButtonStyle.gray,
+    )
+    async def dismiss(
+        self, interaction: discord.Interaction, _: discord.ui.Button
+    ) -> None:
+        assert not is_dm(interaction.user)
+        if interaction.user.id == self._message.author.id or is_mod(interaction.user):
+            for reply in message_to_codeblocks[self._message]:
+                await reply.delete()
+            return
+
+        await interaction.response.send_message(
+            "You can't dismiss this message.", ephemeral=True
+        )
+
+    @discord.ui.button(
+        label="Freeze",
+        emoji="❄️",
+        style=discord.ButtonStyle.gray,
+    )
+    async def freeze(
+        self, interaction: discord.Interaction, button: discord.ui.Button
+    ) -> None:
+        assert not is_dm(interaction.user)
+        if interaction.user.id == self._message.author.id or is_mod(interaction.user):
+            frozen_messages.add(self._message)
+            button.disabled = True
+            await interaction.response.edit_message(view=self)
+            await interaction.followup.send(
+                "Message frozen. I will no longer react to"
+                " what happens to your original message.",
+                ephemeral=True,
+            )
+        else:
+            await interaction.response.send_message(
+                "You can't freeze this message.", ephemeral=True
+            )
+
+    @discord.ui.button(
+        label="Replace my message",
+        emoji="🔄",
+        style=discord.ButtonStyle.gray,
+    )
+    async def replace(
+        self, interaction: discord.Interaction, _: discord.ui.Button
+    ) -> None:
+        assert interaction.message
```

Comment:
Anyone can replace messages right now.
```suggestion
            assert not is_dm(interaction.user)
        if interaction.user.id != self._message.author.id:
            await interaction.response.send_message(
                "You can't replace this message.", ephemeral=True
            )
            return

            assert interaction.message
```

Reply:
Oh i messed up the indentation.

Reply:
Fixed in #154 (accidentally pushed there instead of here lol).